### ISSUE TITLE
Parsing issues with Subsidiaries

### ISSUE THREAD

**Issue by nullbind** [NetSPI/NetblockTool#3]
**Parsing issues with Subsidiaries**
There a few different errors that occur related to subsid parsing.  Below are some examples.

**Example 1**

```
python3 NetblockTool.py -v "[]REDACTED]"                                                  
  _   _      _   _     _            _    _____           _
 | \ | | ___| |_| |__ | | ___   ___| | _|_   _|__   ___ | |
 |  \| |/ _ \ __| '_ \| |/ _ \ / __| |/ / | |/ _ \ / _ \| |
 | |\  |  __/ |_| |_) | | (_) | (__|   <  | | (_) | (_) | |
 |_| \_|\___|\__|_.__/|_|\___/ \___|_|\_\ |_|\___/ \___/|_|
 
[*] Retrieving networks using Google Dorking for[REDACTED] (usually < 30 pages)
  [*] Status: Scrape complete                    
[*] Retrieving ARIN objects using keyword [REDACTED]*
[*] Processing 133 retrieved ARIN objects
Traceback (most recent call last):est/customer/C05765541                     
  File "/home//[REDACTED]/NetblockTool/NetblockTool.py", line 1003, in get_customer_info
    return_street = parsed['customer']['streetAddress']['line']['$']
TypeError: list indices must be integers or slices, not str

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home//[REDACTED]/NetblockTool/NetblockTool.py", line 2033, in <module>
    results = main(arg_target, arg_company_name, query, arg_verbose, arg_threshold, arg_geo, arg_address, arg_address_grep, arg_version, arg_quiet, arg_poc, arg_no_google)
  File "/home//[REDACTED]/NetblockTool/NetblockTool.py", line 199, in main
    return_cust = get_customer_info(item)
  File "/home//[REDACTED]/NetblockTool/NetblockTool.py", line 1012, in get_customer_info
    address += ', '+str(parsed['customer']['streetAddress']['line'][i]['$']).rstrip()
KeyError: '$'
```

**Example 2**

```
OS: WSL Kali & Kali 2021.1
Command:  python3 NetblockTool.py -v "[REDACTED]" -s -sn "[REDACTED]" -sp
Error: 
`
  _   _      _   _     _            _    _____           _
 | \ | | ___| |_| |__ | | ___   ___| | _|_   _|__   ___ | |
 |  \| |/ _ \ __| '_ \| |/ _ \ / __| |/ / | |/ _ \ / _ \| |
 | |\  |  __/ |_| |_) | | (_) | (__|   <  | | (_) | (_) | |
 |_| \_|\___|\__|_.__/|_|\___/ \___|_|\_\ |_|\___/ \___/|_|
[*] Getting subsidiary information for [REDACTED]
  [*] Gathering company information for [REDACTED] from EDGAR database
Traceback (most recent call last):
  File "/home/[REDACTED]/Downloads/NetblockTool/NetblockTool.py", line 2080, in <module>
    for result in get_subsidiaries(arg_subsid_name, arg_verbose, arg_subsid_alt, arg_quiet):
  File "/home/[REDACTED]/Downloads/NetblockTool/NetblockTool.py", line 1487, in get_subsidiaries
    temp.append(edgar.Edgar().getCikByCompanyName(company))
  File "/home/[REDACTED]/.local/lib/python3.9/site-packages/edgar/edgar.py", line 31, in __init__
    all_companies_array[i] = (item_arr[0], item_arr[1])
IndexError: list index out of range`
Using the the tool without the subsidiaries flags works as intended.

```

**Comment by silosw** [comment]
Hi,

The second issue seems to be a result of rate limiting. I did some digging into the referenced file, edgar.py.

When rate limiting isn't affecting us, variable item_arr takes on the form of a list composed by what I think is the subsidiary name, an ID number, and an empty string.

`item_arr = ['A-CNO-21-FUND, A SERIES OF ANGELLIST-NOAL-FUNDS, LLC', '0001684484', '']`

When we're being rate limited, the composition of variable item_arr's list becomes malformed. Below are three examples of what item_arr gets initialized to when the loop in class Edgar's constructor runs. 

**len(item_arr) = 2**
`item_arr = ['<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http', '//www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">']`

**len(item_arr) = 2**
`item_arr = ['<html xmlns="http', '//www.w3.org/1999/xhtml">']`

**len(item_arr) = 1**
`item_arr = ['<head>']`

The following line of code is where everything goes wrong. 
`all_companies_array[i] = (item_arr[0], item_arr[1])`

This operation is creating a tuple out of the first two elements found in variable item_arr. It works fine so long as len(item_arr) >= 2. When it reaches the third instance above, where len(item_arr) = 1, it tries to index into item_arr[1] which doesn't exist. That's why we get the error "IndexError: list index out of range."

I was able to fix this particular issue by modifying the constructor for class Edgar to the following. 

```
    def __init__(self):
        rate_limited = 1
        while rate_limited:
            all_companies_page = requests.get("https://www.sec.gov/Archives/edgar/cik-lookup-data.txt")
            all_companies_content = all_companies_page.content.decode("latin1")
            all_companies_array = all_companies_content.split("\n")
            item_arr = all_companies_array[0].split(":")
            if item_arr[0] != '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http':
                rate_limited = 0

        del all_companies_array[-1]
        all_companies_array_rev = []
        for i, item in enumerate(all_companies_array):
            item_arr = []
            if item == "":
                continue
            item_arr = item.split(":")
            all_companies_array[i] = (item_arr[0], item_arr[1])
            all_companies_array_rev.append((item_arr[1], item_arr[0]))
        self.all_companies_dict = dict(all_companies_array)
        self.all_companies_dict_rev = dict(all_companies_array_rev)

```

This patch just retries the request until we're assured that we haven't visited the rate-limited SEC site.

**Comment by venominfosec** [comment]
The SEC now requires all tools to provide identifying information as the value of the User-Agent header, as stated on the following page: https://www.sec.gov/os/accessing-edgar-data

The dependency that NetblockTool uses (edgar) does not yet support adding this header with a custom value. As such, the subsidiary option for NetblockTool is broken until edgar supports this functionality. I will look into adding this header value as a user supplied option for the edgar package.

**Comment by eabase** [comment]
So can't we skip using that outdated dependency then?

**Comment by eabase** [comment]
@venominfosec 
We should be able to override the Edgar requests headers. 
Any idea how to do that?

**Comment by venominfosec** [comment]
This issue is fixed in commit b7cc9909084c93f57f955150f099c8ac7bec23fa
